protos/memcache/plugin.go

```diff
@@ -0,0 +1,379 @@
+	"github.com/elastic/libbeat/common/streambuf"
+	"github.com/elastic/libbeat/logp"
+
+	"github.com/elastic/packetbeat/config"
+	"github.com/elastic/packetbeat/protos"
+)
+
+// Called to initialize the Plugin
+func (mc *Memcache) Init(testMode bool, results chan common.MapStr) error {
+	debug("init memcache plugin")
+	return mc.InitWithConfig(
+		config.ConfigSingleton.Protocols.Memcache,
+		testMode,
+		results,
+	)
+}
+
+// Called to return the configured ports
+func (mc *Memcache) GetPorts() []int {
+	return mc.Ports.Ports
+}
+
+func (mc *Memcache) ParseUdp(pkt *protos.Packet) {
+	defer logp.Recover("ParseMemcache(UDP) exception")
+
+	buffer := streambuf.NewFixed(pkt.Payload)
+
+	requestId, _ := buffer.NetUint16()
+	seqNumber, _ := buffer.NetUint16()
+	numDatagrams, _ := buffer.NetUint16()
+	buffer.Advance(2) // ignore reserved
+	if buffer.Failed() {
+		debug("parsing memcache udp header failed")
+		return
+	}
+
+	debug("udp datagram requestId=%s, seqNumber=%s, numDatagrams=%s",
+		requestId, seqNumber, numDatagrams)
+
+	// find connection object based on ips and ports (forward->reverse connection)
+	tuple := pkt.Tuple
+	connection := mc.udpConnections[tuple.Hashable()]
+	dir := applayer.NetOriginalDirection
+	if connection == nil {
+		connection = mc.udpConnections[tuple.RevHashable()]
+		if connection != nil {
+			dir = applayer.NetReverseDirection
+		}
+	}
+	if connection == nil {
+		connection = &udpConnection{
+			tuple:    pkt.Tuple,
+			memcache: mc,
+		}
+	}
+
+	// get udp transaction combining forward/reverse direction 'streams'
+	// for current requestId
+	trans := connection.transactions[requestId]
+	if trans != nil && trans.timer != nil {
+		stopped := trans.timer.Stop()
+		if stopped {
+			logp.Warn("timer stopped while processing transaction -> create new transaction")
+			trans = nil
+		}
+	}
+	if trans == nil {
+		trans = &udpTransaction{
+			requestId:   requestId,
+			connection:  connection,
+			transaction: nil,
+		}
+	}
+
+	// get UDP transaction stream combining datagram packets in transaction
+	udpMsg := trans.messages[dir]
+	if udpMsg == nil {
+		udpMsg = &udpMessage{
+			numDatagrams: numDatagrams,
+			count:        0,
+		}
+		trans.messages[dir] = udpMsg
+		if numDatagrams > 1 {
+			udpMsg.datagrams = make([][]byte, numDatagrams)
+		}
+	}
+
+	if udpMsg.numDatagrams != numDatagrams {
+		logp.Warn("number of datagram mismatches in stream")
+		connection.killTransaction(trans)
+		return
+	}
+
+	// try to combine datagrams into complete memcached message
+	parsePacket := udpMsg.numDatagrams == 1
+	if !parsePacket && udpMsg.count < udpMsg.numDatagrams {
+		if udpMsg.datagrams[seqNumber] == nil {
+			udpMsg.datagrams[seqNumber] = buffer.Bytes()
+			udpMsg.count++
+		}
+
+		if udpMsg.count == udpMsg.numDatagrams {
+			parsePacket = true
+			buffer = streambuf.New(nil)
+			for _, payload := range udpMsg.datagrams {
+				buffer.Append(payload)
+			}
+			udpMsg.datagrams = nil
+			parsePacket = true
+		}
+	}
```

Comment:
TODO: add mode for ignoring the data part (up to 1MB) by trying to parse first datargram + continue until command was fully read.